lib/common/acl.c

```diff
@@ -798,4 +805,342 @@ crm_acl_get_set_user(xmlNode *request, const char *field, const char *peer_user)
+            ns = !pcmk__check_acl(i_node->parent,
+                                 (const char *) i_node->name,
+                                 xpf_acl_read)
+                 ? NS_DENIED
+                 : !pcmk__check_acl(i_node,
+                                   (const char *) i_node->name,
+                                   xpf_acl_write)
+                   ? NS_READABLE
+                   : NS_WRITABLE;
+            if (ns == NS_WRITABLE) {
+                if (ns_recycle_writable == NULL) {
+                    ns_recycle_writable = xmlNewNs(i_node, NS_WRITABLE,
+                                                   ACL_NS_Q_WRITABLE);
+                }
+                xmlSetNs(i_node, ns_recycle_writable);
+                ret |= PCMK_ACL_VERDICT_WRITABLE;
+            } else if (ns == NS_READABLE) {
+                if (ns_recycle_readable == NULL) {
+                    ns_recycle_readable = xmlNewNs(i_node, NS_READABLE,
+                                                   ACL_NS_Q_READABLE);
+                }
+                xmlSetNs(i_node, ns_recycle_readable);
+                ret |= PCMK_ACL_VERDICT_READABLE;
+            } else if (ns == NS_DENIED) {
+                if (ns_recycle_denied == NULL) {
+                    ns_recycle_denied = xmlNewNs(i_node, NS_DENIED,
+                                                 ACL_NS_Q_DENIED);
+                }
+                xmlSetNs(i_node, ns_recycle_denied);
+                ret |= PCMK_ACL_VERDICT_DENIED;
+            }
+            break;
+        default:
+            break;
+        }
+
+        if (i_node->children != NULL) {
+            ret |= pcmk__eval_acl_as_namespaces_2(i_node->children);
+        }
+    }
+
+    return ret;
+}
+
+int
+pcmk_acl_evaled_as_namespaces(enum pcmk_acl_cred_type cred_type,
+                              const char *cred, xmlDoc *cib_doc,
+                              xmlDoc **acl_evaled_doc)
+{
+    /* XXX requires prior crm_xml_init */
+
+    int ret, version;
+    xmlNode *target, *comment;
+    char comment_buf[256] = "ACLs as evaluated for user ";
+    const char *validation;
+
+    CRM_CHECK(cred != NULL, return -1);
+    CRM_CHECK(cib_doc != NULL, return -1);
+    CRM_CHECK(acl_evaled_doc != NULL, return -1);
+    /* XXX no proper support for groups yet */
+    CRM_CHECK(cred_type == PCMK_ACL_CRED_USER, return -2);
+
+    if (!pcmk_acl_required(cred)) {
+        /* nothing to evaluate */
+        return 0;
+    }
+
+    /* XXX see the comment for this function, pacemaker-4.0 may need
+           updating respectively in the future */
+    validation = crm_element_value(xmlDocGetRootElement(cib_doc),
+                                   XML_ATTR_VALIDATION);
+    version = get_schema_version(validation);
+    if (get_schema_version("pacemaker-2.0") > version
+            || (-1 != get_schema_version("pacemaker-4.0")
+                    && get_schema_version("pacemaker-4.0") <= version)) {
```

Comment:
Let's define a constant for the upper bound

Reply:
For a future unfolding from users-only view.

Reply:
That was meant as a response to `PCMK_ACL_CRED_USER`.
This point is resolved now.